Add Fragment as a named export to React

clemmy opened the pull request:
This PR is a part of the bigger change to introduce `<></>` fragment syntax to JSX. With this named export, we'll be able to de-sugar `<></>` into `React.createElement(React.Fragment)` calls.

Some concerns:

- Not sure if the way I imported dependencies in `ReactFragment-test.js` is the best way to go about things. I initially tried requiring them all at the top, but I got the error `Invariant Violation: EventPluginRegistry: Cannot inject event plugin ordering more than once. You are likely trying to load more than one copy of React.`, and this was my way of getting around it.
- Is there documentation somewhere that I should update for this PR?

acdlite commented on src/isomorphic/modern/class/__tests__/ReactFragment-test.js:
```diff
@@ -0,0 +1,113 @@
+    );
+  });
+
+  it('should render via native renderer', () => {
```
Just the noop renderer is fine :D This doesn't touch any renderer-specific code.

clemmy replied:
Should have said so sooner! :'(

acdlite replied:
I’m sorry :( Didn’t know you were struggling with this. If something’s not renderer specific, we usually just test with the noop renderer. Except for older tests, which use React DOM.

clemmy replied:
Gotcha, good to know! :)

acdlite commented:
> Is there documentation somewhere that I should update for this PR?

Let's hold off on documentation until we're ready to release it. Also shouldn't merge this until after 16.

gaearon commented:
>With this named export, we'll be able to de-sugar <></> into React.createElement(React.Fragment)

Doesn’t it mean we’ll create two fibers for every fragment? One for the wrapper and one for the fragment itself. Seems unfortunate.

acdlite commented:
@gaearon Yes, but I think this is fine for demo purposes and experimenting at FB. We can optimize it later.

acdlite commented:
Although @clemmy if you want to do it the "right" way and avoid the extra fiber, I'm happy to merge that, too :D

NE-SmallTown commented:
> Relevant quip: https://fburl.com/j3hcevus
The link 404 :astonished: 

IIRC, "<></> fragment syntax to JSX" is still under discussion, or you mean it will be accepted to JSX2.0?

acdlite commented:
> Doesn’t it mean we’ll create two fibers for every fragment

Thought this was the case but then I remembered we don’t create an extra Fragment fiber for an array returned from render. For the same reason we don’t create an extra Fragment fiber for the children of a host component. It just becomes that fiber’s child set. So, we could turn `React.Fragment` into a Fragment fiber instead of a FunctionalComponent fiber, but it would still just be one fiber.

acdlite commented:
Could still be worthwhile for other reasons though

gaearon commented:
>So, we could turn React.Fragment into a Fragment fiber instead of a FunctionalComponent fiber, but it would still just be one fiber.

This is what I meant. But I agree it doesn't block this PR.

clemmy commented:
@NE-SmallTown Oops! The link I posted basically just mentions some quirks regarding the `<></>` syntax such as not allowing attributes. Currently, we're experimenting with the syntax support for fragments, but not the rest of the proposals in the JSX 2.0 issue.

trueadm commented:
I though `ReactFragment` would export a string of `#` instead of a functional component?

sophiebits commented:
Or `'#fragment'`? Then to special case it we don't need to somehow detect if it's the same function, which seems treacherous.

clemmy commented:
@trueadm I believe that's more of an implementation detail? The reason I'm pushing for this named export to be merged in ASAP is to save a bit of headache when rolling out fragment syntax changes. I'd be happy to change it to `'#fragment'` after.

gaearon commented:
Would it be hard to implement `#fragment` in the scope of this PR? I'd rather not expose it as a function at all since then changing its type is technically a breaking change.

gaearon commented:
The branching on string type [happens here](https://github.com/facebook/react/blob/18506240c678b41857c2bf523c19af470e09e49c/src/renderers/shared/fiber/ReactFiber.js#L355-L363). We could check for `#fragment` there and create a Fiber with a private `Fragment` type instead that would be implemented as a function.

trueadm commented:
@clemmy It is an implementation detail, but an important one in my opinion. Babel could instead compile to `React.createElement('#fragment', ...)` instead. There are a bunch of reasons for this:

- We're not changing the type from function to string.
- Far easier to determine this a fragment rather than a functional component.
- It's very common for React codebases to set the Babel JSX pragma to `createElement` and rather than import React in their module, they do `import {createElement, Component} from 'React'`. This would force them to go back and add `Fragment`.
- It means the JSX has a fixed coupling with `React` or another pragma needs to be added to avoid this coupling.

gaearon commented:
>The reason I'm pushing for this named export to be merged in ASAP is to save a bit of headache when rolling out fragment syntax changes. 

Could you elaborate on this? How would the export help for the syntax? My understanding is that desugaring JSX to `React.Fragment` is inherently problematic because it requires a second pragma:

```js
<>hi</>
// How do we know it's React.Fragment and not something else?
React.createElement(React.Fragment, null, 'hi')
```

Whereas if we use a string, we don’t have this problem.

```js
<>hi</>
// Works with any pragma.
React.createElement('#fragment', null, 'hi')
```

Why is merging the export helpful in this case? I thought that, if we went with the `'#fragment'` proposal, we’d only need `React.Fragment` as a fallback for people who can’t use the syntax yet (e.g. due to waiting for tools). I don’t see why the export would be useful for the syntax itself.

aweary commented:
cc @developit, since JSX fragments should be easily utilized by the other major libraries using JSX such as Preact.

acdlite commented:
I say we merge this as-is and decide on the implementation details later. We want to start trying out the fragment syntax at Facebook ASAP.

acdlite commented:
I like the `#fragment` idea but I'd also want to hear from other libraries before committing to it. So let's just go with the simplest thing for now. We can put it behind a feature flag maybe if we need to hide it from the public.

gaearon commented:
If we only need to add it to try it at FB, we could hack it into the internal `React-fb.js` entry point instead. Just like we did with `createClass` during migration.

gaearon commented:
But yes, either that, or let’s put it behind a feature flag. I don’t want this to become a public API before we’re truly committed to it.

acdlite commented:
I'm fine with hacking it onto the export, too. I didn't realize we were still debating whether to export `React.Fragment`, only the underlying implementation.

trueadm commented:
@acdlite I don't see this as the simplest thing

trueadm commented:
Opps sorry about that, my phone went a bit crazy and closed the PR. :/

gaearon commented:
@acdlite 

I agree exporting `React.Fragment` is uncontroversial but the fact that it's a function means somebody can call it directly. Now changing it to a string is technically a breaking change.

I understand it's not likely somebody will do it with a new API but we'd be technically breaking semver when we change it to export `'#fragment'`. I’d like to avoid that since it‘s easily avoidable by doing it on FB side.

acdlite commented:
Meh, everything change is breaking if you dig enough. I'm not too worried about someone calling it :D

acdlite commented:
Even if go with `#fragment` (which I like), I'm still in favor of the export for non-JSX users and for adding keys.

acdlite commented:
so aside from the breaking change concern, I view this discussion as implementation bikeshedding.

acdlite commented:
@clemmy How about you put this behind a feature flag? Then we can merge it without worrying about breakages.

trueadm commented:
@acdlite I think there's some confusion here. The export would still exist, but it would be `module.exports = '#fragment'` rather than exporting a functional component. So non-JSX users can still use it, but it's also a non-breaking API change.

acdlite commented:
Yeah I get that, I'm just not sure I count that as a breaking change. Why would someone call `React.Fragment()` directly? We would document it an opaque component type.

trueadm commented:
@acdlite You've put it on a public API and it has type data to say it's a function – why wouldn't someone try and call it directly, I'd never assume it would be a component because it hasn't got `Component` in the name.

clemmy commented:
@gaearon I'm saying it saves a bit of a headache because then I wouldn't have to worry about www using a fork of React which has something for `<></>` to de-sugar to, which doesn't necessarily mean exporting `React.Fragment`, but just exporting something and having it synced into www.

I don't have strong opinions on either `'#fragment'` or `React.Fragment`. All up for `'#fragment'` if it's not going to an easy change, with better long term benefits.

Also, what's the internal `React-fb.js` entry point?

gaearon commented:
>Also, what's the internal React-fb.js entry point?

If you open www on nuclide and type that, you'll see a file that internal `React` is really aliased to. So you can safely add something there instead of the open source React if you just need to get something working for www.

clemmy commented:
@gaearon Ok, sounds good! I'll do that option for now instead. I'll reserve this PR for when things are a bit more finalized, where React will export `'#fragment'` and do the special case handling.

StevenZiu commented:
any update?

clemmy commented:
Updated PR with `React.Fragment` resolving to `'#fragment'`. 

There are a couple of points of discussion I want to bring up:

- ReactFragment no longer shows up in DevTools, is that behavior what we want?
- Inside `/src/renderers/shared/fiber/ReactFiber.js` line 359, I use a string literal `'#fragment'` instead of importing `ReactFragment` from `ReactBasesClasses` because the import didn't seem to work. This causes a coupling between `ReactFiber.js` and the string used for `ReactFragment`.
- I'm not sure if there's currently key support for `Fragment` fibers right now. Currently, in this PR, setting a key on the Fragment export is a noop.

acdlite commented on src/isomorphic/modern/class/ReactBaseClasses.js:
```diff
@@ -163,8 +163,11 @@ asyncComponentPrototype.render = function() {
   return this.props.children;
 };
 
+const ReactFragment = '#fragment';
```
Let's just move this to `ReactEntry.js`, since it's not a class anymore.

acdlite commented on src/isomorphic/modern/element/__tests__/ReactJSXElementValidator-test.js:
```diff
@@ -109,6 +109,19 @@ describe('ReactJSXElementValidator', () => {
+  it('does not warns for fragments of multiple elements without keys', () => {
+    spyOn(console, 'error');
+
+    void (
```
Let's actually render these elements. The missing key warning is fired during reconciliation, not element creation.

acdlite commented on src/renderers/shared/fiber/ReactFiber.js:
```diff
@@ -348,18 +343,24 @@ exports.createFiberFromText = function(
 function createFiberFromElementType(
-  type: mixed,
-  key: null | string,
+  element: ReactElement,
```
Seems like there's no point in this being a separate function anymore. Let's inline it into `createFiberFromElement`.

acdlite commented:
Also, please rebase this on master to get rid of the conflicts.

acdlite commented:
> ReactFragment no longer shows up in DevTools, is that behavior what we want?

What does it look like? Screenshot?

> However, we need to be able to key fragments.

Please add that in this PR, along with a test. The key is the second argument to `createFiber` (right now you're passing null). You may also need to update ReactChildFiber.

acdlite commented:
These are the lines you'll need to change in ReactChildFiber to add support for keyed fragments.

https://github.com/facebook/react/blob/8ac9600574dc4b4445b32b2d7bcba3c9c0bed6f1/src/renderers/shared/fiber/ReactChildFiber.js#L669-L670

https://github.com/facebook/react/blob/8ac9600574dc4b4445b32b2d7bcba3c9c0bed6f1/src/renderers/shared/fiber/ReactChildFiber.js#L763

clemmy commented:
DevTool:
<img width="298" alt="screen shot 2017-10-12 at 1 00 35 pm" src="https://user-images.githubusercontent.com/3696934/31516445-732083b2-af4d-11e7-8721-b249cfdf3f30.png">

JSX:
![screen shot 2017-10-12 at 1 01 01 pm](https://user-images.githubusercontent.com/3696934/31516446-7337e5fc-af4d-11e7-8f5d-0d2a8d01dc1c.png)

trueadm commented:
Great work!

I'm happy with the string and also quite like @sebmarkbage idea about using a symbol and then falling back to a string if not available (similar to `$$typeof` for `ReactElement`). 

Also not sure if you saw, but these's also a conflict too :)

clemmy commented:
@trueadm Sounds good :) Currently blocked on this PR because I'm not sure what behavior we want for reconciliation between `React.Fragment`'s and implicit child sets. According to @acdlite, they should semantically be the same (e.g. an un-keyed `React.Fragment` should be the semantically the same as any implicit childset: `<React.Fragment><div /></React.Fragment>` and `<div />` are both semantically equivalent in that both childsets don't have a key, but discussion is still ongoing.

clemmy commented:
I used the word `semanticcally` 3 times in one comment. 😆

clemmy commented:
Conclusions reached:

- top level keyed `React.Fragment` => create fragment fiber
- top level un-keyed `React.Fragment` => no fragment fiber created
- nested fragments => fragment fiber created

thysultan commented:
Does `#fragment: string` vs `Fragment: function/class` affect the possible use case of officially using `React.Fragment` as a way of delivering component chunks; Assuming component boundaries are the units used to divide work when cooperative scheduling is enabled? i.e

```js
<ul>
	<React.Fragment>
		<li>1</li>
		<li>2</li>
		<li>3</li>
	</React.Fragment>

	<React.Fragment>
		<li>3</li>
		<li>4</li>
		<li>5</li>
	</React.Fragment>

	<React.Fragment>
		<li>6</li>
		<li>7</li>
		<li>8</li>
	</React.Fragment>
</ul>
```

clemmy commented:
PR isn't complete yet, I need the export to be a symbol and to rebase onto master.

clemmy commented:
@thysultan There's no change to using `React.Fragment`. You can see its usage in the tests:
https://github.com/clemmy/react/blob/6aa2c905a096c27fee883fc6a675b4c4753bc1d6/src/isomorphic/modern/class/__tests__/ReactFragment-test.js

acdlite commented on src/isomorphic/modern/class/__tests__/ReactFragment-test.js:
```diff
@@ -0,0 +1,180 @@
+    ReactNoop.render(element);
+    ReactNoop.flush();
+
+    expect(ReactNoop.getChildren()).toEqual([
```
Let's use `div` and `span` helper functions like we do in the other test modules. https://github.com/facebook/react/blob/56e5288b042523729d5f8b84a3e569b0b96a329a/src/renderers/shared/fiber/__tests__/ReactIncrementalSideEffects-test.js#L26-L33

acdlite commented on src/isomorphic/modern/class/__tests__/ReactFragment-test.js:
```diff
@@ -0,0 +1,180 @@
+    );
+    ReactNoop.flush();
+
+    var instanceB = instance;
```
If Stateful never re-renders, then instance points to the previous value. So we need to test that Stateful actually re-renders to avoid a false _negative_ ;) The `ops.push()` pattern is useful here. Could also check if `componentDidUpdate` is called, rather than comparing instances.

acdlite commented on src/isomorphic/modern/class/__tests__/ReactFragment-test.js:
```diff
@@ -0,0 +1,180 @@
+    expect(instanceB).toBe(instanceA);
+  });
+
+  it('should not preserve state when switching to a nested fragment', function() {
```
Same feedback as above

acdlite commented on src/isomorphic/modern/class/__tests__/ReactFragment-test.js:
```diff
@@ -0,0 +1,180 @@
+    expect(instanceB).not.toBe(instanceA);
+  });
+
+  it('should preserve state in a reorder', function() {
```
and again

acdlite commented on src/isomorphic/modern/element/__tests__/ReactJSXElementValidator-test.js:
```diff
@@ -101,19 +103,34 @@ describe('ReactJSXElementValidator', () => {
+  it('does not warns for fragments of multiple elements without keys', () => {
+    spyOn(console, 'error');
+
+    if (ReactDOMFeatureFlags.useFiber) {
```
Looks like your PR branch has fallen behind master. This flag no longer exists since we deleted Stack. You'll want to rebase it.

acdlite commented on src/renderers/shared/fiber/ReactChildFiber.js:
```diff
@@ -993,7 +1009,7 @@ function ChildReconciler(shouldClone, shouldTrackSideEffects) {
       ;
       oldFiber !== null && !step.done;
-      newIdx++, (step = newChildren.next())
+      newIdx++, step = newChildren.next()
```
What's with all these style changes? Prettier version mismatch?

acdlite commented on src/renderers/shared/fiber/ReactChildFiber.js:
```diff
@@ -371,7 +372,22 @@ function ChildReconciler(shouldClone, shouldTrackSideEffects) {
     priority: PriorityLevel,
   ): Fiber {
-    if (current === null || current.type !== element.type) {
+    if (
```
It looks like the only difference between this branch and the next is that if it's a Fragment fiber, the `pendingProps` should be the children. Can we merge these two branches and only branch on that assignment?

acdlite commented on src/renderers/shared/fiber/ReactChildFiber.js:
```diff
@@ -1155,6 +1171,17 @@ function ChildReconciler(shouldClone, shouldTrackSideEffects) {
             existing._debugOwner = element._owner;
           }
           return existing;
+        } else if (child.tag === Fragment && element.type === '#fragment') {
```
Same as previous comment

acdlite commented on src/renderers/shared/fiber/ReactChildFiber.js:
```diff
@@ -1304,6 +1331,30 @@ function ChildReconciler(shouldClone, shouldTrackSideEffects) {
+          if (
+            newChild.type === '#fragment' &&
+            newChild.key === null &&
+            (returnFiber.tag === HostRoot ||
```
Why are you checking the type of return fiber? Does it matter?

acdlite commented on src/renderers/shared/fiber/ReactFiber.js:
```diff
@@ -399,11 +360,42 @@ function createFiberFromElementType(
+  priorityLevel: PriorityLevel,
+): Fiber {
+  // TODO: Consider supporting keyed fragments. Technically, we accidentally
+  // support that in the existing React.
```
Do we? We might have in Stack, but that's no longer the "existing React." :D If we don't support it in 16, let's not worry about it.

clemmy replied:
This was an existing TODO comment in the code :D. If it's no longer relevant, should I remove it?

sophiebits commented:
@thysultan React can already pause work in between each of the `<li>` elements in your example; adding fragment wrappers doesn't help it do that.

gaearon commented on src/isomorphic/modern/element/__tests__/ReactJSXElementValidator-test.js:
```diff
@@ -99,19 +100,49 @@ describe('ReactJSXElementValidator', () => {
     );
   });
 
+  it('does not warns for fragments of multiple elements without keys', () => {
```
Typo: warn

clemmy replied:
😆 was matching the other typos in this file. I'll fix the typo for all the tests.

gaearon commented on src/isomorphic/modern/element/__tests__/ReactJSXElementValidator-test.js:
```diff
@@ -99,19 +100,49 @@ describe('ReactJSXElementValidator', () => {
+      </React.Fragment>,
+    );
+
+    expectDev(console.error.calls.count()).toBe(0);
```
This is unnecessary as long as you remove both this *and* the `spyOn` call.
We assume tests shouldn't warn by default unless you `spyOn`.

gaearon commented on src/isomorphic/modern/element/__tests__/ReactJSXElementValidator-test.js:
```diff
@@ -99,19 +100,49 @@ describe('ReactJSXElementValidator', () => {
+
+    expectDev(console.error.calls.count()).toBe(1);
+    expectDev(console.error.calls.argsFor(0)[0]).toContain(
+      'Encountered two children with the same key',
```
Worth including the key into the message? To make sure we don't mess it up when printing.

gaearon commented on src/isomorphic/modern/element/__tests__/ReactJSXElementValidator-test.js:
```diff
@@ -156,15 +187,20 @@ describe('ReactJSXElementValidator', () => {
     spyOn(console, 'error');
 
-    void <Component><Component /><Component /></Component>;
+    void (
```
Let's change to `TestUtils.renderIntoDocument` too for consistency?

clemmy replied:
Will do :) The reason I didn't add `TestUtils.renderIntoDocument` here was because it's only necessary to fire key warnings, which this test wasn't testing.

gaearon commented on src/renderers/shared/fiber/ReactChildFiber.js:
```diff
@@ -1361,6 +1412,18 @@ function ChildReconciler(shouldClone, shouldTrackSideEffects) {
     if (isObject) {
       switch (newChild.$$typeof) {
         case REACT_ELEMENT_TYPE:
+          // is this right?
```
Did this get resolved?

sebmarkbage commented on src/isomorphic/ReactEntry.js:
```diff
@@ -27,6 +27,12 @@ if (__DEV__) {
+  (typeof Symbol === 'function' &&
+    Symbol.for &&
+    Symbol.for('react.fragment')) ||
+  0xad9c;
```
For elements we picked `0xeac7` because it kind of looks like `React`. What does this look like? :)

The other ones have used increasing numbers since.

https://github.com/facebook/react/blob/d9c1dbd61772f8f8ab0cdf389e70463d704c480b/packages/react-reconciler/src/isomorphic/ReactCoroutine.js#L23-L24

https://github.com/facebook/react/blob/d9c1dbd61772f8f8ab0cdf389e70463d704c480b/packages/react-reconciler/src/isomorphic/ReactPortal.js#L19

The next one would be `0xeacb`

clemmy replied:
Wasn't aware of that, it looks like the number `44444` if you convert it to decimal. 😆

sebmarkbage commented on src/isomorphic/classic/element/ReactElementValidator.js:
```diff
@@ -19,6 +19,12 @@
+    Symbol.for &&
+    Symbol.for('react.fragment')) ||
+  0xad9c;
+
```
We should move this into the `__DEV__` block because I don't think it gets DCE otherwise since `Symbol.for()` could be a stateful function.

clemmy replied:
Is `ReactElementValidator` not used when `__DEV__` is off?

sebmarkbage replied:
Nope.

sebmarkbage commented on src/isomorphic/classic/element/ReactElementValidator.js:
```diff
@@ -229,7 +235,10 @@ function validatePropTypes(element) {
+    var validType =
+      typeof type === 'string' ||
+      typeof type === 'function' ||
+      type === REACT_FRAGMENT_TYPE;
```
This should probably just allow any symbol and numbers so that we can use this space for other built-in fun things. E.g. passing it to host configs.

sebmarkbage commented on src/renderers/shared/fiber/ReactChildFiber.js:
```diff
@@ -523,6 +530,18 @@ function ChildReconciler(shouldClone, shouldTrackSideEffects) {
         case REACT_ELEMENT_TYPE: {
+          if (newChild.type === REACT_FRAGMENT_TYPE) {
+            const created = createFiberFromFragment(
+              isArray(newChild.props.children)
```
What if it is `<>{new Set()}</>`? That is an iterable item. In all other cases arrays and iterables are treated the same. I believe this changes the semantics because you nest it one extra time.

Who do you need to do this at all? If the children set of a fragment is a single item, it should fall into the branch that optimizes for reconciling a single child. The is same as passing a single child to a Host component.

So I don't think you should need to wrapper at all.

clemmy replied:
@sebmarkbage I'm having a tough time with this part. From my understanding, there's only a branch that optimizes for reconciling a single child in `reconcileSingleElement`. However, if we want to preserve the semantics of not having a fragment fiber in a top level array, then we need to reconcile using `reconcileChildrenArray`, which only takes in an array of children.

clemmy replied:
Actually, I can probably recurse on `reconcileChildFibers`. Does that make sense? I'll test that right now.

sebmarkbage replied:
You can't recurse on reconcileChildFibers because that will flatten a nested set of single element fragments which is not what we want.

You can however fork it and if it's another fragment go for the multi-child case.

Then you can try to unify them and share code.

sebmarkbage commented on src/renderers/shared/fiber/ReactChildFiber.js:
```diff
@@ -616,6 +636,17 @@ function ChildReconciler(shouldClone, shouldTrackSideEffects) {
+              return updateFragment(
+                returnFiber,
+                oldFiber,
+                isArray(newChild.props.children)
```
Same thing.

sebmarkbage commented on src/renderers/shared/fiber/ReactChildFiber.js:
```diff
@@ -712,6 +749,17 @@ function ChildReconciler(shouldClone, shouldTrackSideEffects) {
+            return updateFragment(
+              returnFiber,
+              matchedFiber,
+              isArray(newChild.props.children)
```
Same thing.

sebmarkbage commented on src/renderers/shared/fiber/ReactChildFiber.js:
```diff
@@ -1207,7 +1256,9 @@ function ChildReconciler(shouldClone, shouldTrackSideEffects) {
           existing.ref = coerceRef(child, element);
-          existing.pendingProps = element.props;
+          existing.pendingProps = element.type === REACT_FRAGMENT_TYPE
+            ? element.props.children
```
Unlike the other branches, this doesn't do the wrapper thing? (It's good that it doesn't but inconsistent. Probably indicates that there is a missing test.)

sebmarkbage replied:
This unifies both behind `child.type === element.type`. This is neat but it also means that we're sometimes comparing strings to numbers/symbols. That could potentially deopt the type hints in this function. It might actually be better to check the `child.tag` to ensure that it is a `HostComponent` or `Fragment` and compare the type in separate branches.

In fact, that might already be a bug incase we ever add strings to `fiber.type` on other fibers than HostComponent.

sebmarkbage commented on src/renderers/shared/fiber/ReactChildFiber.js:
```diff
@@ -1361,6 +1412,17 @@ function ChildReconciler(shouldClone, shouldTrackSideEffects) {
+            return reconcileChildrenArray(
+              returnFiber,
+              currentFirstChild,
+              isArray(newChild.props.children)
```
Same thing.

sebmarkbage commented on src/renderers/shared/fiber/ReactFiber.js:
```diff
@@ -304,62 +310,27 @@ exports.createFiberFromElement = function(
+    fiber.pendingProps = element.props;
+  } else if (type === REACT_FRAGMENT_TYPE) {
+    fiber = createFiberFromFragment(
+      Array.isArray(element.props.children)
```
Same thing.

sebmarkbage commented on src/renderers/shared/fiber/ReactFiber.js:
```diff
@@ -398,10 +370,42 @@ function createFiberFromElementType(
+  key: null | string,
+): Fiber {
+  const fiber = createFiber(Fragment, key, internalContextTag);
+  fiber.type = REACT_FRAGMENT_TYPE;
```
If you always compare against `.tag` instead of `.type` you can get rid of this assignment. We shouldn't unnecessarily use fields we don't really need and for the fragment tag this field is unnecessary.

clemmy replied:
Would it be preferable to have a branch on `current.tag === Fragment && element.type === REACT_FRAGMENT_TYPE` at places where it would usually be a `current.type === element.type`? That the initial way I implemented this but @acdlite suggested I can add this field to not have to special case.

sebmarkbage replied:
yea.

clemmy commented:
Addressed your comments @sebmarkbage . 

For branching on tag type before we compare `element.type` and `oldFiber.type` during `updateElement`, `reconcileSingleElement`, and possible other places, I made a followup issue at https://github.com/facebook/react/issues/11290. 

Also, I might have scope creeped a bit in this PR by allowing symbol and number types inside the validator for `createElement`. Let me know if I should remove it from this PR. 

Gonna merge master into this branch to fix conflicts and it should be good to go. :)

sebmarkbage commented on packages/react-reconciler/src/ReactChildFiber.js:
```diff
@@ -375,27 +380,34 @@ function ChildReconciler(shouldClone, shouldTrackSideEffects) {
+    if (
+      current !== null &&
+      (current.type === element.type ||
+        (current.tag === Fragment && element.type === REACT_FRAGMENT_TYPE))
```
```js
(current.tag === Fragment ? element.type === REACT_FRAGMENT_TYPE : current.type === element.type)
```
That way you don't need to compare the unused `type` field on fragments.

sebmarkbage commented on packages/react-reconciler/src/ReactChildFiber.js:
```diff
@@ -1203,11 +1238,17 @@ function ChildReconciler(shouldClone, shouldTrackSideEffects) {
+        // TODO: Split these into branches based on typeof type
+        if (
+          child.type === element.type ||
+          (child.tag === Fragment && element.type === REACT_FRAGMENT_TYPE)
```
Same thing.
```js
(current.tag === Fragment ? element.type === REACT_FRAGMENT_TYPE : current.type === element.type)
```

sebmarkbage commented on packages/react-reconciler/src/ReactChildFiber.js:
```diff
@@ -1342,9 +1383,140 @@ function ChildReconciler(shouldClone, shouldTrackSideEffects) {
+      );
+    }
+
+    if (isArray(newChild)) {
```
Doesn't this mean that `return <>{[...]}</>` is treated the same as `return <>{...}</>`. That's not the same as `return [...]` vs `return [[...]]`. It flattens an extra level because reconcileChildrenArray avoids the extra fragment but in this case there should always be a fragment.

We should probably have a test for this.

sebmarkbage replied:
I think I gave you bad advice. The function you wanted to fork was `updateSlot`.

clemmy replied:
Bringing this up made me think about some other permutations of fragment and array combinations given our current desired behavior of top-level fragments behaving the same as a top-level array. I jotted some of them down here (with the ones I'm unsure about marked with `?`:
https://gist.github.com/clemmy/b3ef00f9507909429d8aa0d3ee4f986b
Does that look right to you?

sebmarkbage commented on packages/react-reconciler/src/ReactChildFiber.js:
```diff
@@ -1361,6 +1533,17 @@ function ChildReconciler(shouldClone, shouldTrackSideEffects) {
+          // so that it is treated as a set of children
+          // Otherwise, we follow the normal flow.
+          if (newChild.type === REACT_FRAGMENT_TYPE && newChild.key === null) {
+            return reconcileChildFibersForTopLevelFragment(
```
I actually think you don't even need to fork anything. You just need to call `updateSlot` here instead.

sebmarkbage commented:
Actually as for ReactChildren I guess you could make an argument that this should work like elements (terminal) and not arrays because we model them as elements. But these also feel like arrays.

Either switching between `<></>` and `[]` is unsafe, or switching between `<></>` and `<div></div>` is unsafe.

However in every other way we've treated these as replacement for arrays so it seems like ReactChildren should treat them as such. However if we ever extend them to have other properties like event handlers, then treating them as arrays would be lossy.

sebmarkbage commented:
We also need to add this to the server renderer which currently just uses React.Children.toArray but if that doesn't drill deeply then it might try to renderer the element. Needs some tests.

sebmarkbage commented:
There is a completely different approach we could consider too. Just changing React.createElement to special case this and return an array.

clemmy commented:
Conclusions from offline discussion with @sebmarkbage :
- `React.Children` will treat fragments as a terminal element
- In this gist (https://gist.github.com/clemmy/b3ef00f9507909429d8aa0d3ee4f986b), switching between arrays and un-keyed fragments will preserve state (e.g. Child will be preserved when rendering `[<><Child /></>]` and then `<>[<Child />]</>`)

uniqueiniquity commented on packages/react/src/React.js:
```diff
@@ -25,6 +25,12 @@ if (__DEV__) {
+  (typeof Symbol === 'function' &&
+    Symbol.for &&
+    Symbol.for('react.fragment')) ||
+  0xeacb;
```
@clemmy Why is the fallback here a number instead of a string? Not sure I follow.
I'm curious so I can accurately update the react typings for TypeScript, and it seems weird to have the first argument to createElement be a string or a symbol or a number.

uniqueiniquity replied:
(sorry, didn't mean to add myself as a reviewer)

clemmy replied:
The fallback here is a couple of reasons:

- Strings are currently used for specifying host components in `React.createElement`, it would be a bit confusing to special case fragments into this space. See https://github.com/facebook/react/pull/10783#discussion_r145575197
- Number comparison is faster than string comparison in JS

uniqueiniquity replied:
Sounds good. Thanks!

gaearon commented on packages/react/src/ReactElementValidator.js:
```diff
@@ -228,7 +228,11 @@ function validatePropTypes(element) {
+    var validType =
+      typeof type === 'string' ||
+      typeof type === 'function' ||
+      typeof type === 'symbol' ||
```
What happens when it's an unrecognised symbol? Do we throw later?

gaearon replied:
Same question for number.

clemmy replied:
Currently there's nothing implemented to throw later, since @sebmarkbage suggested leaving this space open for 
 fun things (https://github.com/facebook/react/pull/10783#discussion_r145575197)

gaearon replied:
What does the user see though? We should at least warn because if it silently does nothing, then making it do something could be a breaking change (if someone starts relying on it by accident).

clemmy replied:
So the behavior we want is to warn if it's not equal to `REACT_FRAGMENT_TYPE`?

clemmy commented:
The behavior now matches with what I describe in this [gist](https://gist.github.com/clemmy/b3ef00f9507909429d8aa0d3ee4f986b). I still need to update server rendering tests before this PR is ready. :)

gaearon commented on packages/react-dom/src/server/ReactPartialRenderer.js:
```diff
@@ -483,6 +489,12 @@ class ReactDOMServerRenderer {
   constructor(children: mixed, makeStaticMarkup: boolean) {
     var flatChildren;
+    if (
+      React.isValidElement(children) &&
```
If this is false then we're doing this check twice now. Maybe restructure so we don't have to?

clemmy replied:
In order to preserve the optimization on the if/else, I'd end up with something like:
```
    if (React.isValidElement(children)) {
      const element = ((children: any): ReactElement);
      if (element.type === REACT_FRAGMENT_TYPE) {
        if (React.isValidElement(element.props.children)) {
          flatChildren = [element.props.children];
        } else {
          flatChildren = toArray(element.props.children);
        }
      } else {
        flatChildren = toArray(element);
      }
    } else {
      flatChildren = toArray(children);
    }
```
The code is a bit more confusing, but it does save a duplicate check. What do you think?

gaearon replied:
Can you extract this in a function with early returns?

```js

    function flattenTopLevelChildren(children: mixed): FlatReactChildren {
      if (!React.isValidElement(children)) {
        return toArray(children);
      }
      const element = ((children: any): ReactElement);
      if (element.type !== REACT_FRAGMENT_TYPE) {
        return [element];
      }
      const fragmentChildren = element.props.children;
      if (!React.isValidElement(fragmentChildren)) {
        return toArray(fragmentChildren);
      }
      const fragmentChildElement = ((fragmentChildren: any): ReactElement);
      return [fragmentChildElement];
    }
```

clemmy replied:
This looks very nice. 👍

gaearon commented on packages/react-dom/src/server/ReactPartialRenderer.js:
```diff
@@ -570,13 +582,9 @@ class ReactDOMServerRenderer {
       if (nextChild === null || nextChild === false) {
         return '';
       } else {
-        if (React.isValidElement(nextChild)) {
```
Maybe flatten this to "else if" at the level above? Now that we have else's in both branches.

clemmy replied:
Oops! Somehow missed this haha.

clemmy commented on packages/react-reconciler/src/ReactChildFiber.js:
```diff
@@ -385,17 +388,8 @@ function ChildReconciler(shouldClone, shouldTrackSideEffects) {
-      created.return = returnFiber;
-      return created;
-    } else {
+    // TODO: Split these into branches based on typeof type
```
https://github.com/facebook/react/issues/11290 It's to ensure correctness between type checking when comparing across functions, classes, strings, and potentially numbers/symbols.

sebmarkbage replied:
This should split based on `tag`, not `typeof`. It doesn't really matter if you're going to check the type anyway. Some of the other duplicates of this comment is confusing since in those cases you already do that branching.

clemmy replied:
In the duplicates of this comment, I only do branching on `tag === REACT_FRAGMENT_TYPE`. My understanding was that we wanted to branch in such a way so that we only compare types of the same type (`function`, `string`, `number`, `symbol`, etc.). Perhaps I misunderstood.

sebmarkbage commented on packages/react-reconciler/src/ReactChildFiber.js:
```diff
@@ -533,14 +539,25 @@ function ChildReconciler(shouldClone, shouldTrackSideEffects) {
+              returnFiber.internalContextTag,
+              expirationTime,
+              newChild.key,
+            );
```
I wonder if we should have a warning for `ref` on a fragment, either in ReactChildFiber or in ReactElementValidator?

clemmy replied:
Oops, I assumed that `ref` would appear in `element.props`. I think it makes sense to do the check in ReactElementValidator since that's the earliest possible time we can catch the potential bug right? What are the advantages of having the warning inside ReactChildFiber?

sebmarkbage replied:
I think other ref warnings are in the reconciler. The benefit of that is that we (or others) can create changes to the reconciler (such as adding ref support on fragments) without changing the isomorphic package. Another issue is that people can create elements in other ways (e.g. by inlining) and not get the warning.

But as you said, it's nice to get it earlier for stack traces.

I don't really have a preference.

clemmy replied:
When you say that people can create elements by inlining, this can only happen in non-dev mode right? (I'm assuming that when you mention inlining, you are referring to this transform: https://github.com/babel/babel/tree/master/packages/babel-plugin-transform-react-inline-elements)

sebmarkbage commented on packages/react-dom/src/server/ReactPartialRenderer.js:
```diff
@@ -206,6 +212,22 @@ function getNonChildrenInnerMarkup(props) {
+    return toArray(children);
+  }
+  const element = ((children: any): ReactElement);
+  if (element.type !== REACT_FRAGMENT_TYPE) {
```
This is just an optimization right? Not semantic meaning. I wonder if this is just slowing down the normal case where this is `true` since the check will happen again then.

clemmy replied:
Is there a way to do less checks? I think that every way we re-write this, there still ends up being 3 checks. https://github.com/facebook/react/pull/10783/files/48bcaa7198ef5d97bfc78badd1636a836a076cda..27312d0ebbc3ebbdaa8dc48e6f70119847267389#r147533483

blling commented:
Is there any docs about the usecase of this pr?

clemmy commented:
@blling There will be docs about fragments when this is officially announced. But basically, this PR provides an export `React.Fragment` for JSX `<></>` to de-sugar to.

blling commented:
What is `<></>` used for, i have never used it, could you give some details ?

sophiebits commented:
It desugars to a React.Fragment element. We'll publish docs when this is ready.

blling commented:
OK, thanks.

jvitoroc commented:
Is there a collision with [`Fragments`](https://reactjs.org/docs/react-component.html#fragments)?

jvitoroc commented:
Is there a name collision with https://reactjs.org/docs/react-component.html#fragments ?

gaearon commented:
This is a more convenient API for the same feature. We'll unify the docs later.

gaearon commented on packages/react/src/ReactElementValidator.js:
```diff
@@ -226,9 +236,44 @@ function validatePropTypes(element) {
+function validateFragmentProps(fragment) {
+  currentlyValidatingElement = fragment;
+
+  for (const key of Object.keys(fragment.props)) {
```
We shouldn't have used this. It introduces reliance on `Symbol` which we don't currently require.

I think we should add a lint against `for ... of` as it's easy to use accidentally and it's not obvious what it compiles down to.

xgqfrms-gildata commented:
good to go!

xgqfrms-gildata commented:
![image](https://user-images.githubusercontent.com/29531779/37317102-ee512222-269d-11e8-80fc-cabefe91eb76.png)

I have some questions, anybody can explain it more in detail?

https://reactjs.org/docs/react-component.html#fragments

1. ...fragments, which don’t require keys for `static items`:

> is that mean we should use a unique key for each `dynamic items`?

2. just like the below codes show, `li` no need a unique key.

```jsx
render() {
  return (
    <React.Fragment>
      <li>First item</li>
      <li>Second item</li>
      <li>Third item</li>
    </React.Fragment>
  );
}

```

> is that mean `React.Fragment` will auto give a unique key for each `li` item?

```html

    <React.Fragment>
      <li key="A">First item</li>
      <li key="B">Second item</li>
      <li key="C">Third item</li>
    </React.Fragment>

```

clemmy commented:
@xgqfrms-gildata No keys are required in `React.Fragment`s because they are static in nature. However, keys are required when using the array syntax.

The wording in the documentation is a bit convoluted since the array syntax used to be called fragments. To make it clearer:

Instead of
```
Don't forget to add keys to elements in a fragment to avoid the key warning
```

It should be

```
Don't forget to add keys to elements in an array to avoid the key warning
```

xgqfrms-gildata commented:
@clemmy Thanks for your time.

> good explanation.
